dockstore-webservice/src/main/java/io/dockstore/webservice/languages/WDLHandler.java

```diff
@@ -120,25 +125,41 @@ public boolean isValidWorkflow(String content) {
             }
 
             for (String importPath : importPaths) {
-                SourceFile importFile = new SourceFile();
+                boolean toAdd = true;
+                for (String path : alreadyImported) {
+                    if (Objects.equals(path, importPath)) {
+                        toAdd = false;
+                        break;
+                    }
+                }
 
-                final String fileResponse = sourceCodeRepoInterface.readGitRepositoryFile(fileType, version, importPath);
-                if (fileResponse == null) {
-                    SourceCodeRepoInterface.LOG.error("Could not read: " + importPath);
-                    continue;
+                if (toAdd) {
+                    SourceFile importFile = new SourceFile();
+
+                    final String fileResponse = sourceCodeRepoInterface.readGitRepositoryFile(fileType, version, importPath);
+                    if (fileResponse == null) {
+                        SourceCodeRepoInterface.LOG.error("Could not read: " + importPath);
+                        continue;
+                    }
+                    importFile.setContent(fileResponse);
+                    importFile.setPath(importPath);
+                    importFile.setType(SourceFile.FileType.DOCKSTORE_WDL);
+                    imports.put(importFile.getPath(), importFile);
                 }
-                importFile.setContent(fileResponse);
-                importFile.setPath(importPath);
-                importFile.setType(SourceFile.FileType.DOCKSTORE_WDL);
-                imports.put(importFile.getPath(), importFile);
             }
 
+            importPaths.addAll(alreadyImported);
+
             for (SourceFile sf : imports.values()) {
-                imports.putAll(processImports(sf.getContent(), version, sourceCodeRepoInterface));
+                imports.putAll(processImports(sf.getContent(), version, sourceCodeRepoInterface, importPaths));
             }
         } catch (IOException e) {
             throw new CustomWebApplicationException("Internal server error, out of space",
                 HttpStatus.SC_INSUFFICIENT_SPACE_ON_RESOURCE);
+        } finally {
+            if (tempDesc != null) {
```

Comment:
Part of the point of using `deleteQuietly` is you don't actually need a null check. 
`file - file or directory to delete, can be null`